### ISSUE TITLE
Better self transition SVG rendering (and the ability to render internal self transitions)

### ISSUE THREAD

**Issue by LeeDoty** [sverweij/state-machine-cat#95]
### Context
I'm trying to push SMCAT to render SVGs of our in-house high TPS nested state engine.
I love the idea of SMCAT and it's exactly what I need (simple way to visualize with no attempts at behavior)

### Proposed Behavior
When I render any self transition, it probably should be a segment of an elipse... a simple arc, and multiple self transitions should probably appear concentrically around each other.  
Also, room needed for these transitions should make space on the graph (internally or expanding the bounds)
And finally, there should be a way to render internal transitions, especially internal self transitions.  These are generally represented as traversing internally in a state, (inside the boundaries as opposed to outside), specifically, the internal self transitions should render as described above: concentric simple arcs, just inside the host state boundary instead of outside.

### Current Behavior
2 problems:
1. when I render self transitions, especially multiple self transitions (say for different events) the graph becomes unreadable and it actually looks like error output (though it's not, all the event lines are just tangled)
2. there doesn't seem to be a way to mark a transition as internal in SMCAT format and scxml markup doesn't render differently.

To be clear, I'm suggesting that it's an error to render a self transition (internal or external) as a complex path.

### Considered alternatives
I tried multiple renderers in your online formatter (I typically use the atom plugin for prototyping... kudos again, BTW) dot was the only legible one, and the others were completely unreadable.
I can render SMCAT and SCXML formats, but I can't use SCXML, as the newer name aliasing you added (kudos again... I need this to simplify nested state names) doesn't work for SCXML inputs, and SCXML renderer doesn't help anyway as the internal transitions don't render.

Here's an example SMCAT that shows the tangled behavior of self transitions when rendered to SVG:  (note: the interenal/external transitions are tangled left and above their host state B)

```
NestedStateTests [label="NestedStateTests"] : 
    /entry Root Entry
    /exit Root Exit
{
    NestedStateTests.A [label="A"] : 
    {
        NestedStateTests.A.A1 [label="A1"] : 
            /entry A1 Entry
            /exit A1 Exit
        {
            NestedStateTests.A.A1 => NestedStateTests.A.A2 : I1
            /A1->A2 Action;
        },
        NestedStateTests.A.A2 [label="A2"] : 
            /entry A2 Entry
            /exit A2 Exit
        {
            NestedStateTests.A.A2 => NestedStateTests.B.B2 : I1
            /A2->B2 Action;
        },
        NestedStateTests.A.A3 [label="A3"] : ;
        NestedStateTests.A => NestedStateTests.B.B2 : I2
        /A->B2 Action;
        NestedStateTests.A => NestedStateTests.B : E1
        /A->B Action;
        NestedStateTests.A => NestedStateTests.C : E2
        /A->C Action;
    },
    NestedStateTests.B [label="B"] : 
        /entry B Entry
        /exit B Exit
    {
        NestedStateTests.B.B2 [label="B2"] : 
            /entry B2 Entry
            /exit B2 Exit,
        NestedStateTests.B.B1 [label="B1"] : 
            /entry B1 Entry
            /exit B1 Exit;
        NestedStateTests.B => NestedStateTests.A : E1;
        NestedStateTests.B => NestedStateTests.B : E3
        /B->B (external) Action;
        NestedStateTests.B => NestedStateTests.B : <<internal>> E4
        /B->B (internal) Action;
    },
    NestedStateTests.C [label="C"] : 
        /entry C Entry
        /exit C Exit
    {
        NestedStateTests.C.C1 [label="C1"] : 
            /entry C1 Entry
            /exit C1 Exit
        {
            NestedStateTests.C.C1 => NestedStateTests.C.C2 : --AUTO--
            /C1 -> C2 Action;
        },
        NestedStateTests.C.C2 [label="C2"] : 
            /entry C2 Entry
            /exit C2 Exit;
    };
};
```

**Comment by sverweij** [comment]
hi @LeeDoty thanks for raising this issue. Or actually two, if I understand correctly

## self transitions of composite states that are part of a composite state look ugly

😬 You're right. That's unacceptably ugly - and unreadable when there's > 1 transition. I'll see if I can massage GraphViz to play nice with these...

```
outer {
    inner.one,
    inner.two {
        innerst => innerst: 
          innerst => innerst
          (sorta ok);
    };
    inner.one => inner.one:
      inner.one => inner.one
      (sorta ok);
    inner.two => inner.two [color="red"]:
      inner.two => inner.two
      (sweet mother!!);
};
outer => outer: 
  outer => outer
  (sorta ok);
```

![inner-outer](https://user-images.githubusercontent.com/4822597/70379934-218b6a00-1934-11ea-9f40-ea3408eb97d7.png)



## cannot distinguish internal and external self transitions

Thanks for bringing that up. It's currently part of the language nor of the internal data model, but it should totally be added. 

I'll try to keep close to what's in the [SCXML standard](https://www.w3.org/TR/scxml/#transition) and ...
- add a `type` attribute to transitions in the internal data model - when omitted it's assumed to be _external_
- heed the `type` attribute in scxml
- add a `type` attribute to the smcat language
- render `type=internal` different from 

```
outer {
  inner;
};

outer => outer: external transition; // default is type=external, just like in SCXML
outer => outer [type=internal]: internal transition;
```

![untitled](https://user-images.githubusercontent.com/4822597/70379874-f2c0c400-1932-11ea-9d52-2752fabc3631.png)

**Comment by sverweij** [comment]
Hi @LeeDoty I've opened a PR to add at least the distinction between internal and external self transitions (☝️ #96). It now works for SCXML and JSON inputs - you can have a peek in the [beta deployment](https://sverweij.gitlab.io/state-machine-cat/) on gitlab.

If you use e.g. this SCXML 

```scxml
<?xml version="1.0" encoding="UTF-8"?>
<scxml xmlns="http://www.w3.org/2005/07/scxml" version="1.0">
    <state id="outer">
        <state id="inner">
        </state>
        <transition event="external_transition" target="outer"/> <!-- external is assumed-->
        <transition event="internal_transition" type="internal" target="outer"/>
    </state>
</scxml>
```

... over there it'll come out ~ like so:

![state-machine-2019-12-07T21_46_35 157Z](https://user-images.githubusercontent.com/4822597/70381034-20facf80-1944-11ea-803b-b11b7dd0ca2a.png)


> The addition to the smcat language will follow suit - I'll ping in this thread.
> I'm looking into the ugly self-renders as well - but that'll need (at least) a good night's sleep first ...

**Comment by sverweij** [comment]
I've identified two slight improvements to make the self transitions on states that are at the same time... 
- ... nested in another state 
- and composite ...
look a bit better.

First one is in PR #97 - that's just published to [gitlab as a beta deployment](https://sverweij.gitlab.io/state-machine-cat/)

<img width="478" alt="image" src="https://user-images.githubusercontent.com/4822597/70831142-0c11b680-1df2-11ea-8421-b2489f515359.png">

**Comment by LeeDoty** [comment]
Excellent!!  Thanks so much for the quick action.

Currently I'm using the online visualizer and the atom plugin.  Is there a beta version of either, or would I need to use the beta code command line renderer?

again, thanks!  I'll further thank you with yet another bug report....

**Comment by sverweij** [comment]
I've pushed the changes to npm as _latest_ - I'll update the atom plugin as well. Should be on atom.io within a few minutes.

> Multiple external transitions still look gnarly. I have a solution candidate for that, but it needs
> some more intensive code juggling. I'll ping in this issue if it's ready to be reviewed.

**Comment by sverweij** [comment]
@LeeDoty the multiple external transition problem on external states is fixed as well - it's not the nicest looking, though. I consider it acceptable for now because (1) they're now distinguishable (2) I have the impression multiple external self-transitions on compound states are sort of an edge case (but I'd be glad to be disavowed of that impression).

As I'm writing this it's being deployed on [https://sverweij.gitlab.io/state-machine-cat/](https://sverweij.gitlab.io/state-machine-cat/), so when you read it you should be able to check. Feedback is welcome! 

I'll merge & release to the _latest_ channels later this week.

**Comment by sverweij** [comment]
It's published as 5.4.1 on npm and state-machine-cat.js.org and as 4.6.7 on atom.io. (Feedback still welcome).
